doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,571 @@
+- **Note**: Moved earlier to enable per-file frame and target code generation
+
+#### 13. **TargetCodeGenerationPass**
+- **Purpose**: Generate both frame code and target-specific code for each SDFG file by calling the appropriate target code generators
+- **Input**: Split SDFGs with all previous analysis
+- **Output**: pipeline_results["code_objects"] = List[CodeObject] with complete code
+- **Current Location**: Combined from `DaCeCodeGenerator.generate_code()` and target-specific `get_generated_codeobjects()`
+- **Note**: This pass calls the target code generators (CppCodeGen, GPUCodeGen, FPGACodeGen, etc.) to generate platform-specific code
+
+#### 14. **HeaderGenerationPass**
+- **Purpose**: Generate C/C++ header files for SDFG interface
+- **Input**: CodeObjects with complete code
+- **Output**: pipeline_results["headers"] = {call_header, sample_main}
+- **Current Location**: `generate_headers()`, `generate_dummy()`
+
+## Information Flow Design
+
+### Main Code Generation Pipeline
+
+```python
+class CodeGenerationPipeline(Pipeline):
+    """Complete code generation pipeline for DaCe SDFGs."""
+
+    def __init__(self):
+        super().__init__([
+            # Phase 1: Analysis
+            TypeInferencePass(),
+            LibraryExpansionPass(),
+            MetadataCollectionPass(),
+            AllocationAnalysisPass(),
+            ControlFlowAnalysisPass(),
+            TargetAnalysisPass(),
+
+            # Phase 2: Transformations
+            AllocationPass(),
+            CopyToMapPass(),
+            ConditionalPipeline([
+                (lambda r: 'cuda' in r.get('targets', []), StreamAssignmentPass()),
+                (lambda r: 'fpga' in r.get('targets', []), FPGAPreprocessingPass()),
+            ]),
+            TaskletLanguageLoweringPass(),
+
+            # Phase 3: Code Generation
+            StateStructCreationPass(),
+            MemletLoweringPass(),
+            SDFGSplittingPass(),
+            FrameAndTargetCodeGenerationPass(),
+            HeaderGenerationPass(),
+        ])
+```
+
+### Target-Specific Sub-Pipelines
+
+```python
+class CUDACodegenPipeline(Pipeline):
+    """CUDA-specific code generation pipeline."""
+
+    def __init__(self):
+        super().__init__([
+            CUDATargetValidationPass(),
+            CUDAMemoryAnalysisPass(),
+            CUDAKernelExtractionPass(),
+            CUDAStreamAssignmentPass(),
+            CUDAKernelGenerationPass(),
+            CUDALaunchCodePass(),
+        ])
+
+class FPGACodegenPipeline(Pipeline):
+    """FPGA-specific code generation pipeline."""
+
+    def __init__(self):
+        super().__init__([
+            FPGAResourceAnalysisPass(),
+            FPGAStreamingPass(),
+            FPGAHLSGenerationPass(),
+            FPGAHostCodePass(),
+        ])
+```
+
+## Information Flow and Reuse
+
+### Pipeline Results Schema
+
+The `pipeline_results` dictionary will contain structured information to maximize reuse:
+
+```python
+pipeline_results = {
```

Comment:
Should the design document include a more detailed description of the (sub-)dictionary keys and example values?

Reply:
Too many details that are design-irrelevant. Removed.